Initial support for multi-networking

dotnwat opened the pull request:
This PR adds the plumbing for supporting multi-network configuration. The basic idea is that listeners and advertised apis are associated using common naming. Given the following configuration:

```
  kafka_api:
    - name: public
      address: "0.0.0.0"
      port: 9092
    - name: private
      address: "0.0.0.0"
      port: 9093

  advertised_kafka_api:
    - name: public
      address: public.asdf.com
      port: 9092
    - name: private
      address: internal.network
      port: 9093
```

kafka will report public.asdf.com as an advertised listener if the client is connected on the public listener connection, and similar behavior for the private network.

senior7515 commented on src/v/config/property.h:
```diff
@@ -80,11 +80,37 @@ class property : public base_property {
+ * schemas that have simplified special cases.
+ */
+template<typename T>
+class one_or_many_property : public property<std::vector<T>> {
```
wow. hot!!!

senior7515 commented on src/v/rpc/server.cc:
```diff
@@ -131,8 +136,8 @@ ss::future<> server::stop() {
     ss::sstring proto_name = _proto ? _proto->name() : "protocol not set";
     vlog(
       rpclog.info, "{} - Stopping {} listeners", proto_name, _listeners.size());
-    for (auto&& l : _listeners) {
```
sounds like `&&` was a bug?

dotnwat replied:
no i don't think it is a bug. it's just the value type is known and it looks like there nothing to std::forward

senior7515 commented on src/v/rpc/types.h:
```diff
@@ -177,8 +177,20 @@ inline result<T> get_ctx_data(result<client_context<T>>&& ctx) {
+    ss::socket_address addr;
+
+    explicit server_endpoint(ss::socket_address addr)
+      : addr(addr) {}
```
i think you should use forward ctors via 

``` ..: name("default"), addr(addr) {}```

senior7515 commented on src/v/kafka/requests/metadata_request.cc:
```diff
@@ -412,17 +412,19 @@ ss::future<response_ptr> metadata_api::process(
-                  .port = b->kafka_api_address().port(),
-                  .rack = b->rack()};
-            });
+          for (const auto& broker : brokers) {
```
very cool!

senior7515 commented on src/v/config/tests/advertised_kafka_api_test.cc:
```diff
@@ -34,6 +34,28 @@ static auto const advertised_kafka_api_conf = "  advertised_kafka_api:\n"
+    "    - address: 192.168.1.1\n"
+    "      port: 9999\n"
+    "  seed_servers:\n"
+    "    - host: \n"
```
i don't see the test for the multiple `name, addr, port`

dotnwat replied:
I'll add that. The existing test converts to vector, but of size one

senior7515 reviewed (approved):
Really quite fantastic!!! 

WHat are your thoughts on defaulting the name to "default"

dotnwat commented:
> Really quite fantastic!!!
> 
> WHat are your thoughts on defaulting the name to "default"

We could _display_ empty stings in logs specially? I don't really care.

mmaslankaprv commented:
This looks really great, the only thing I miss is the backward compatibility for `raft::configuration` it also stores `model::broker`

BenPope reviewed (approved):
I was trying to test it with the helm chart, but had no luck.

dotnwat commented:
- tests for multiple endpoints in broker object
- support backwards compatibility for raft::group_configuration
- use constructor delegation

> I was trying to test it with the helm chart, but had no luck.

offline chatting looks like this is an issue in the helm chart with constructing the proper config file

dotnwat commented:
> This looks really great, the only thing I miss is the backward compatibility for `raft::configuration` it also stores `model::broker`

done

dotnwat commented:
- tests for multiple endpoints in broker object
- support backwards compatibility for raft::group_configuration
- use constructor delegation

dotnwat commented:
@mmaslankaprv wanna check the backwards compat stuff?

dotnwat commented:
- tests for multiple endpoints in broker object
- support backwards compatibility for raft::group_configuration
- use constructor delegation

https://github.com/vectorizedio/redpanda/compare/631febd7ac7c4ebed82091bbdd32912e70cb34fc..1a7e1d5

@mmaslankaprv if you have time it'd be great to look at (1) the backward compat i added and(2) the test i added for that feature.

senior7515 commented on src/v/model/metadata.h:
```diff
@@ -219,9 +242,39 @@ struct topic_metadata {
+ * Old version for use in backwards compatibility serialization /
+ * deserialization helpers.
+ */
+struct broker_v0 {
```
thanks for doing this.

senior7515 reviewed (approved):
lgtm.
